keylime-agent/src/config.rs

```diff
@@ -59,244 +59,228 @@ pub static DEFAULT_AGENT_DATA_PATH: &str = "agent_data.json";
-            (
-                "contact_ip".to_string(),
-                if let Some(ref s) = self.agent.contact_ip {
-                    s.to_string().into()
-                } else {
-                    "".into()
-                },
-            ),
-            (
-                "contact_port".to_string(),
-                if let Some(ref s) = self.agent.contact_port {
-                    s.to_string().into()
-                } else {
-                    "".into()
-                },
-            ),
-            (
-                "registrar_ip".to_string(),
-                self.agent.registrar_ip.to_string().into(),
-            ),
-            (
-                "registrar_port".to_string(),
-                self.agent.registrar_port.into(),
-            ),
-            (
-                "enable_agent_mtls".to_string(),
-                self.agent.enable_agent_mtls.into(),
-            ),
-            (
-                "keylime_dir".to_string(),
-                self.agent.keylime_dir.to_string().into(),
-            ),
-            (
-                "server_key".to_string(),
-                if let Some(ref s) = self.agent.server_key {
-                    s.to_string().into()
-                } else {
-                    "".into()
-                },
-            ),
-            (
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub(crate) struct AgentConfig {
+    pub version: Option<String>,
+    pub uuid: Option<String>,
+    pub ip: Option<String>,
+    pub port: Option<u32>,
+    pub contact_ip: Option<String>,
+    pub contact_port: Option<u32>,
+    pub registrar_ip: Option<String>,
+    pub registrar_port: Option<u32>,
+    pub enable_agent_mtls: Option<bool>,
+    pub keylime_dir: Option<String>,
+    pub server_key: Option<String>,
+    pub server_cert: Option<String>,
+    pub server_key_password: Option<String>,
+    pub trusted_client_ca: Option<String>,
+    pub enc_keyname: Option<String>,
+    pub dec_payload_file: Option<String>,
+    pub secure_size: Option<String>,
+    pub tpm_ownerpassword: Option<String>,
+    pub extract_payload_zip: Option<bool>,
+    pub enable_revocation_notifications: Option<bool>,
+    pub revocation_actions_dir: Option<String>,
+    pub revocation_notification_ip: Option<String>,
+    pub revocation_notification_port: Option<u32>,
+    pub revocation_cert: Option<String>,
+    pub revocation_actions: Option<String>,
+    pub payload_script: Option<String>,
+    pub enable_insecure_payload: Option<bool>,
+    pub allow_payload_revocation_actions: Option<bool>,
+    pub tpm_hash_alg: Option<String>,
+    pub tpm_encryption_alg: Option<String>,
+    pub tpm_signing_alg: Option<String>,
+    pub ek_handle: Option<String>,
+    pub run_as: Option<String>,
+    pub agent_data_path: Option<String>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub(crate) struct KeylimeConfig {
+    pub agent: AgentConfig,
+}
+
+impl AgentConfig {
+    pub fn get_map(&self) -> Map<String, Value> {
```

Comment:
OK for me, but maybe we can drop concept of "configuration as a dictionary", and use the struct directly

Reply:
We use the struct directly, the collection to get the map is only used to implement `Source` to allow overriding options with the values obtained from the environment variable.